### PR TITLE
DDF-2180: Provide FTPS support for the FTP Endpoint

### PR DESCRIPTION
**Pull request by peterhuffer**
#### What does this PR do?
Implements FTPS support for the FTP Endpoint
-Provides support for Explicit FTPS
-Provides configuration for 1 and 2 way SSL authentication
-Wait for connections to close before updating server with new configuration 
-Downgrade Apache Mina to 2.0.6. This is because Apache FTPServer has not been updated to newer versions of Apache Mina. This fixes OutOfMemoryError being caught incorrectly, and also gets rid of some erroneous stack traces.
-Moves FTP module from `/catalog/core/ftp` to `/catalog/ftp`
#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged; if a component team is listed, at least one of its members needs to approve)?
@troymohl @mweser @emmberk @lcrosenbu @ricklarsen @stustison 
#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@jaymcnallie
@kcwire
#### How should this be tested?
-Build and run an instance of DDF
-Install `catalog-core-ftp` feature through the Admin UI under DDF Catalog

Insecure connection
-From the command line, `ftp localhost 8021`

1 way authentication
-Ensure `Client Authentication` is set to `Want` in the FTP Endpoint configuration.
-In FileZilla (or another FTPS client of your choice) setup a new site configuration (see screenshots)
-Input password `admin` and accept the Certificate in the pop-up window
-Verify in the FileZilla message log window that TLS was initialized.

2 way authentication
-In FTP Endpoint configuration, change `Client Authentication` from Want to Need.
-Download CyberDuck. CyberDuck FTPS client supports client authentication (FileZilla doesn't), or you can use `curl`. `curl` command is `curl --insecure -v -S --ftp-ssl -ftp-method singlecwd -E /path/to/client/cert:certPassword ftp://admin:admin@localhost:8021`. You _should_ be able to use the auto generated `localhost.p12` cert generated after initial startup in `etc/certs/`
-Accept server certificate in popup
-Select client certificate (needs to be trusted by server, `localhost.p12` should already be trusted) to send to server in next popup. The client certificate should be in your systems keychain manager in its Certificates category.

#### Any background context you want to provide?
FTPS is required for encrypting communications between server and client using SSL/TLS.
#### What are the relevant tickets?
DDF-2180
#### Screenshots (if appropriate)
In FileZilla under File->Site Manager
screen shot 2016-06-14 at 3 25 19 pm

In CyberDuck
screen shot 2016-06-16 at 2 00 38 pm

#### Checklist:
- [x] Documentation Updated
- [x] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

### CODE REVIEW

**Comment by peterhuffer** [comment]
Accidently closed the last one https://github.com/codice/ddf/pull/863

File: distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc [hunk 1]
```diff
@@ -1118,7 +1144,7 @@ After creating a custom Ftplet, it needs to be added to the FTP server’s Ftple
 
 ====== From an FTP client:
 
-The FTP endpoint can be accessed from any FTP client of choice. Some common clients are FileZilla, PuTTY, or the FTP client provided in the terminal. The default port number is *8021*. Valid usernames and password are stored in the `<INSTALL_DIR>/etc/users.properties` file.
+The FTP endpoint can be accessed from any FTP client of choice. Some common clients are FileZilla, PuTTY, or the FTP client provided in the terminal. The default port number is *8021*. Valid usernames and password are stored in the `<INSTALL_DIR>/etc/users.properties` file. If FTPS is enabled with 2 way authentication, a client that supports client authorization is required.
```
Comment (stustison) [inline distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc]: You don't need to tell them that valid usernames/passwords are in users.properties. On a hardened system, this file is removed anyways. Valid users are any users that can log in via our security framework.
Also "2 way authentication" -> "2-way TLS" and "client authorization" -> "client authentication"

File: catalog/core/catalog-core-ftp/src/main/java/ddf/catalog/core/ftp/UserManagerImpl.java [hunk 2]
```diff
@@ -169,9 +176,15 @@ private boolean checkAdmin(String username) {
...
+    /**
+     * The {@code UserManager} interface expects only one admin. The first admin to login receives admin
+     * status for this {@code UserManager}
+     *
+     * @param user {@link User} eligible for admin
+     */
     private void setAdmin(User user) {
         Subject subject = ((FtpUser) user).getSubject();
-        if (subject.hasRole("admin") && StringUtils.isEmpty(this.admin)) {
+        if (StringUtils.isEmpty(this.admin) && subject.hasRole("admin")) {
```
Comment (stustison) [inline catalog/core/catalog-core-ftp/src/main/java/ddf/catalog/core/ftp/UserManagerImpl.java]: Don't hardcode the admin role. Use the system property karaf.local.roles

Comment (peterhuffer) [inline catalog/core/catalog-core-ftp/src/main/java/ddf/catalog/core/ftp/UserManagerImpl.java]: Changed to get roles from karaf.local.roles.

File: distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc (hunk 1 above)
Comment (peterhuffer) [inline distribution/docs/src/main/resources/_contents/_catalog-contents/integrating-catalog-contents.adoc]: Updated with suggestions.

**Comment by peterhuffer** [comment]
As pointed out on the last PR, we should be able to rely on authenticating the user with the user certificate instead of a password, however Apache FtpServer does not implement the required 232 response to a USER command. We are going to assume a valid username/password is required.

**Comment by peterhuffer** [comment]
Refactored FTP module from `catalog/core/ftp` to `catalog/ftp`

**Comment by jaymcnallie** [comment]
Built on clean repo and verified all test steps.  👍

**Comment by mweser** [comment]
👍

**Pull request merged**